### PR TITLE
KafkaSupervisor parallel HTTP requests

### LINKED ISSUE #3197: Support parallel HTTP requests in KafkaIndexTaskClient
KafkaIndexTaskClient currently issues HTTP requests serially - i.e. if it needs to tell 30 tasks to pause, it will issue a request to the first task and wait for a response before issuing a request to the next task. Normally this isn't problematic, but when tasks are unresponsive and the requests get retried this can potentially block for a long time. Ideally requests could be made to all tasks in parallel.

### PR DESCRIPTION
**Pull request by dclim**
The initial version of the supervisor was single-threaded which worked well enough in general but resulted in poor responsiveness when HTTP requests to the index task ChatHandlers didn't return in a timely fashion. The main change in this PR is to issue HTTP requests to multiple tasks simultaneously to improve the responsiveness of the supervisor. The overall coordination logic has not been changed.

Summary of changes:

**KafkaIndexTaskClient**
- add better logging
- make retries and timeouts configurable
- add ListenableFuture async wrapper methods

**KafkaSupervisor**
- change concurrently accessed data structures to concurrent variants
- separate issuing of HTTP requests (through KafkaIndexTaskClient) and supervision logic so that we can send parallel HTTP requests without waiting for responses and then wait on the responses as a group instead of sequentially

**Testing**
- existing tests all pass with minor modifications mainly to support new mock expectations
- added new tests to test async KafkaIndexTaskClient methods; tests run parameterized with single/multiple threads
- added additional tests to KafkaSupervisor to test expected handling of unresponsive tasks and graceful shutdown behavior
- added basic integration test for Kafka indexing service to test task creation, handoff, and proper offset continuation between tasks

Has been running for a few days in our test cluster with no related issues. 

Fixes #3197

### CODE REVIEW

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisorIOConfig.java [hunk 1]
```diff
@@ -54,7 +58,11 @@ public KafkaSupervisorIOConfig(
       @JsonProperty("period") Period period,
       @JsonProperty("useEarliestOffset") Boolean useEarliestOffset,
       @JsonProperty("completionTimeout") Period completionTimeout,
-      @JsonProperty("lateMessageRejectionPeriod") Period lateMessageRejectionPeriod
+      @JsonProperty("lateMessageRejectionPeriod") Period lateMessageRejectionPeriod,
+      @JsonProperty("workerThreads") Integer workerThreads,
+      @JsonProperty("chatThreads") Integer chatThreads,
+      @JsonProperty("chatRetries") Long chatRetries,
+      @JsonProperty("httpTimeout") Period httpTimeout
```
Comment (gianm) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisorIOConfig.java]: [minor] These seem more like tuning config things than i/o config things. i/o is supposed to be things directly related to how we read and write data; tuning is supposed to be all the incidentals.

File: integration-tests/src/test/java/io/druid/tests/indexer/ITKafkaIndexingServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+import java.util.concurrent.Callable;
+
+/*
+ * This is a test for the Kafka indexing service.
+ */
+@Guice(moduleFactory = DruidTestModuleFactory.class)
+public class ITKafkaIndexingServiceTest extends AbstractIndexerTest
```
Comment (gianm) [inline integration-tests/src/test/java/io/druid/tests/indexer/ITKafkaIndexingServiceTest.java]: Sick!

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java [hunk 3]
```diff
@@ -76,33 +83,82 @@ public TaskNotRunnableException(String message)
...
+  )
   {
     this.httpClient = httpClient;
     this.jsonMapper = jsonMapper;
     this.taskInfoProvider = taskInfoProvider;
+    this.httpTimeout = httpTimeout;
+    this.numRetries = numRetries;
     this.retryPolicyFactory = createRetryPolicyFactory();
+
+    String threadId = String.format("KafkaIndexTaskClient-%s", dataSource);
```
Comment (gianm) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java]: `String.format("KafkaIndexTaskClient-%s-%%d", dataSource)` should work too, although changing is unnecessary

File: docs/content/development/extensions-core/kafka-ingestion.md [hunk 4]
```diff
@@ -162,6 +162,10 @@ For Roaring bitmaps:
...
 |`completionTimeout`|ISO8601 Period|The length of time to wait before declaring a publishing task as failed and terminating it. If this is set too low, your tasks may never publish. The publishing clock for a task begins roughly after `taskDuration` elapses.|no (default == PT30M)|
 |`lateMessageRejectionPeriod`|ISO8601 Period|Configure tasks to reject messages with timestamps earlier than this period before the task was created; for example if this is set to `PT1H` and the supervisor creates a task at *2016-01-01T12:00Z*, messages with timestamps earlier than *2016-01-01T11:00Z* will be dropped. This may help prevent concurrency issues if your data stream has late messages and you have multiple pipelines that need to operate on the same segments (e.g. a realtime and a nightly batch ingestion pipeline).|no (default == none)|
+|`workerThreads`|Integer|The number of threads that will be used by the supervisor for asynchronous operations.|no (default == min(10, taskCount))|
+|`chatThreads`|Integer|The number of threads that will be used for communicating with indexing tasks.|no (default == min(10, taskCount * replicas))|
```
Comment (nishantmonu51) [inline docs/content/development/extensions-core/kafka-ingestion.md]: Looks like we will have 20 threads at min for each dataSource we are trying to ingest ? 
The number of threads might become a bottleneck in near future for the Overlord. 
How do you think about having a single async pool configurable via the overlord runtime.props and sharing it across multiple supervisors ?

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java [hunk 5]
```diff
@@ -547,118 +579,167 @@ private void updatePartitionDataFromKafka()
...
+        // and if not add it to taskGroups or pendingCompletionTaskGroups (if status = PUBLISHING)
+        TaskGroup taskGroup = taskGroups.get(taskGroupId);
+        if (!isTaskInPendingCompletionGroups(taskId) && (taskGroup == null || !taskGroup.tasks.containsKey(taskId))) {
+
+          futureTaskIds.add(taskId);
+          futures.add(
+              Futures.transform(
+                  taskClient.getStatusAsync(taskId), new Function<KafkaIndexTask.Status, Boolean>()
+                  {
+                    @Nullable
```
Comment (nishantmonu51) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java]: when will it be null?

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java [hunk 6]
```diff
@@ -757,34 +861,39 @@ private void checkTaskDuration()
...
+        // set endOffsets as the next startOffsets
+        for (Map.Entry<Integer, Long> entry : endOffsets.entrySet()) {
+          partitionGroups.get(groupId).put(entry.getKey(), entry.getValue());
         }
-
-        // remove this task group from the list of current task groups now that it has been handled
-        i.remove();
       }
+
+      // remove this task group from the list of current task groups now that it has been handled
```
Comment (nishantmonu51) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java]: we probably need to handle the case when endOffsets is null, i.e the future threw exception. 
wonder if we should hard kill the task in that case ?

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java [hunk 7]
```diff
@@ -972,13 +1130,15 @@ private void checkCurrentTaskState()
...
-          stopTasksInGroup(taskGroup);
+          futures.add(stopTasksInGroup(taskGroup));
           iTaskGroups.remove();
           break;
         }
       }
       log.debug("Task group [%d] post-pruning: %s", groupId, taskGroup.tasks.keySet());
     }
+
+    Futures.successfulAsList(futures).get(); // wait for all task shutdowns to complete before returning
```
Comment (nishantmonu51) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java]: log warn for failed futures ?

File: docs/content/development/extensions-core/kafka-ingestion.md (hunk 4 above)
Comment (gianm) [inline docs/content/development/extensions-core/kafka-ingestion.md]: IMO this could be a future patch, I think it's out of scope for this particular PR.

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java [hunk 8]
```diff
@@ -76,33 +83,82 @@ public TaskNotRunnableException(String message)
...
 
-  public void stop(String id, boolean publish)
+  public void close()
   {
-    submitRequest(id, HttpMethod.POST, "stop", publish ? "publish=true" : null, true);
+    executorService.shutdown();
   }
 
-  public void resume(String id)
+  public Boolean stop(final String id, final boolean publish)
```
Comment (gianm) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java]: This could be `boolean`, it seems like there's no need for the return value to be nullable.

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java [hunk 9]
```diff
@@ -128,97 +184,251 @@ public void resume(String id)
...
+    catch (NoTaskLocationException e) {
+      return ImmutableMap.of();
+    }
     catch (IOException e) {
       throw Throwables.propagate(e);
     }
   }
 
-  public void setEndOffsets(String id, Map<Integer, Long> endOffsets)
+  public Boolean setEndOffsets(final String id, final Map<Integer, Long> endOffsets)
```
Comment (gianm) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java]: This one also doesn't seem to really need to be nullable.

Comment (gianm) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java]: Ah, I guess you're doing this so you can stuff it in a future? Even so, it'd be nice to have this be `boolean` if that compiles properly. The return value of "Boolean" makes me wonder "Why does this need to be nullable"

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisorIOConfig.java (hunk 1 above)
Comment (gianm) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisorIOConfig.java]: Yeah, this new stuff should be tuning configs.

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java [hunk 10]
```diff
@@ -289,6 +312,7 @@ public void run()
       );
 
       started = true;
+      log.info(spec.toString());
```
Comment (gianm) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java]: `log.info("%s", spec.toString())` is a little safer in case the spec string has some %'s in it. It may be better to combine this with the next line anyway though, since this log line is a bit anemic and context-free.

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java [hunk 11]
```diff
@@ -76,33 +83,82 @@ public TaskNotRunnableException(String message)
...
+
+    String threadId = String.format("KafkaIndexTaskClient-%s", dataSource);
+    this.executorService = MoreExecutors.listeningDecorator(Execs.multiThreaded(numThreads, threadId + "-%d"));
   }
 
-  public void stop(String id, boolean publish)
+  public void close()
   {
-    submitRequest(id, HttpMethod.POST, "stop", publish ? "publish=true" : null, true);
+    executorService.shutdown();
```
Comment (gianm) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java]: Is `shutdownNow` more appropriate (which would interrupt running tasks)?

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java [hunk 12]
```diff
@@ -335,6 +359,8 @@ public void stop(boolean stopGracefully)
         }
         log.info("Shutdown notice handled");
 
+        taskClient.close();
+        workerExec.shutdown();
```
Comment (gianm) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java]: What's the rationale for shutdown vs shutdownNow here?

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java (hunk 5 above)
Comment (gianm) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java]: Guessing this was boilerplate added by IntelliJ; it looks like this won't return null.

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisorIOConfig.java (hunk 1 above)
Comment (dclim) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisorIOConfig.java]: Makes sense, will fix

File: docs/content/development/extensions-core/kafka-ingestion.md (hunk 4 above)
Comment (dclim) [inline docs/content/development/extensions-core/kafka-ingestion.md]: Not 20 threads min but 20 threads max, and that's just for the default which can be overridden if necessary. I considered having a single async pool and actually started implementing it but then thought it was nice to be able to change the concurrency for a supervisor without having to restart the overlord and most users likely won't be running hundreds of supervisors in a single cluster. But definitely open to a shared pool if it feels like the way to go in the future.

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java (hunk 5 above)
Comment (dclim) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java]: yeah just boilerplate, will remove, thanks!

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java (hunk 8 above)
Comment (dclim) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java]: It gets wrapped in a Callable and ListenableFuture in stopAsync() which is why it can't be a primitive.

Comment (dclim) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java]: Actually take it back, it seems to auto-box it, will change.

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java (hunk 9 above)
Comment (dclim) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java]: Sounds good

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java (hunk 12 above)
Comment (dclim) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java]: Hm, I was probably thinking that using shutdown here (and in the task client) would give more chance for any tasks that needed to be stopped/killed to be handled correctly if they didn't respond before the shutdown timeout, but I'm not sure it makes much difference and it's probably more important to ensure the overlord doesn't have any stuck threads hanging around so I agree shutdownNow is better here.

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java (hunk 6 above)
Comment (dclim) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java]: That sounds good to me

File: docs/content/development/extensions-core/kafka-ingestion.md (hunk 4 above)
Comment (nishantmonu51) [inline docs/content/development/extensions-core/kafka-ingestion.md]: agree with @gianm  if we face issues with thread count down the line, we can do that in a separate patch. Not a blocker for this PR.

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java (hunk 7 above)
Comment (nishantmonu51) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java]: sounds good, just wanted to make sure we have some logging in place.

**Comment by gianm** [comment]
👍  LGTM, thanks for the updates @dclim

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java [hunk 13]
```diff
@@ -128,97 +191,251 @@ public void resume(String id)
...
+              "Still waiting for task [%s] to pause; will try again in [%s]",
+              id,
+              new Duration(sleepTime).toString()
           );
           Thread.sleep(sleepTime);
         }
       }
     }
+    catch (NoTaskLocationException e) {
+      return null;
```
Comment (dclim) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java]: @nishantmonu51 cool consistent sounds good, will make the change

Comment (nishantmonu51) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java]: thanks.

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java [hunk 14]
```diff
@@ -335,6 +378,8 @@ public void stop(boolean stopGracefully)
         }
         log.info("Shutdown notice handled");
 
+        taskClient.close();
+        workerExec.shutdownNow();
```
Comment (pjain1) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java]: @dclim In case of graceful shutdown `workerExec` executor service is used to get the highest offset of tasks in the task group, set that as end offset for all tasks and resume them so that they finish and publish. However, since in most cases doing this will take more time than the time out for shutting down supervisor i.e. `SHUTDOWN_TIMEOUT_MILLIS`(15 seconds), `shutdownNow()` will be called on `workerExec`. Thus, when all the futures for pausing tasks resolves, further logic for getting highest offset and resuming them will not run and fail with exception like -
```
Sep 21, 2016 9:42:33 PM com.google.common.util.concurrent.ExecutionList executeListener
SEVERE: RuntimeException while executing runnable com.google.common.util.concurrent.Futures$ChainingListenableFuture@5fde229f with executor com.google.common.
util.concurrent.MoreExecutors$ListeningDecorator@6c5b645
java.util.concurrent.RejectedExecutionException: Task com.google.common.util.concurrent.Futures$ChainingListenableFuture@5fde229f rejected from java.util.conc
urrent.ThreadPoolExecutor@2fde4963[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 10]
        at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
        at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
        at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1369)
        at com.google.common.util.concurrent.MoreExecutors$ListeningDecorator.execute(MoreExecutors.java:484)
        at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
        at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
        at com.google.common.util.concurrent.AbstractFuture.set(AbstractFuture.java:185)
        at com.google.common.util.concurrent.Futures$CombinedFuture.setOneValue(Futures.java:1626)
        at com.google.common.util.concurrent.Futures$CombinedFuture.access$400(Futures.java:1470)
        at com.google.common.util.concurrent.Futures$CombinedFuture$2.run(Futures.java:1548)
        at com.google.common.util.concurrent.MoreExecutors$SameThreadExecutorService.execute(MoreExecutors.java:297)
        at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
        at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
        at com.google.common.util.concurrent.ListenableFutureTask.done(ListenableFutureTask.java:91)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:384)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:251)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```

Comment (dclim) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java]: @pjain1 have you seen this issue while shutting down a supervisor? I considered this possibility but I think that in most cases the shutdown logic should take less than 15 seconds unless the tasks are for some reason not very responsive (and there was an issue around this but #3401 should have fixed it)

Comment (dclim) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java]: @pjain1 I'm going to repost this on #3495 and continue the conversation there
